File: GFG/DP/38.Buy-ans-Sell-Stocks-IV.cpp
// Maximum Profit (K transaction)

// Memoization
class Solution {
public:
    int solve(int idx, int buy, int K, int& N, int A[], vector<vector<vector<int>>>& dp) {
        if(idx==N || K==0) return 0;
        if(dp[idx][buy][K] != -1) return dp[idx][buy][K];
        if(buy) {
            int Buy = -A[idx] + solve(idx+1,0,K,N,A,dp);
            int notbuy = 0 + solve(idx+1,1,K,N,A,dp);
            return dp[idx][buy][K] = max(Buy,notbuy);
        }
        int sell = A[idx] + solve(idx+1,1,K-1,N,A,dp);
        int notsell = 0 + solve(idx+1,0,K,N,A,dp);
        return dp[idx][buy][K] = max(sell,notsell);
    }

    int maxProfit(int K, int N, int A[]) {
        // code here
        vector<vector<vector<int>>> dp(N, vector<vector<int>> (2, vector<int> (K+1,-1)));
        return solve(0,1,K,N,A,dp);
    }
};


// Tabulation
class Solution {

public:

    int maxProfit(int K, int N, int A[]) {
        // code here
        vector<vector<vector<int>>> dp(N+1, vector<vector<int>> (2, vector<int> (K+1,0)));
        for(int idx=N-1; idx>=0; idx--) {
            for(int k=1; k<=K; k++) {
                int Buy = -A[idx] + dp[idx+1][0][k];
                int notbuy = 0 + dp[idx+1][1][k];
                dp[idx][1][k] = max(Buy,notbuy);
                
                int sell = A[idx] + dp[idx+1][1][k-1];
                int notsell = 0 + dp[idx+1][0][k];
                dp[idx][0][k] = max(sell,notsell);
            }
        }
        return dp[0][1][K];
    }
};



// Space Optimization
class Solution {

public:

    int maxProfit(int K, int N, int A[]) {
        // code here
        vector<vector<int>> prev(2, vector<int> (K+1,0));
        vector<vector<int>> curr(2, vector<int> (K+1,0));
        for(int idx=N-1; idx>=0; idx--) {
            for(int k=1; k<=K; k++) {
                int Buy = -A[idx] + prev[0][k];
                int notbuy = 0 + prev[1][k];
                curr[1][k] = max(Buy,notbuy);
                
                int sell = A[idx] + prev[1][k-1];
                int notsell = 0 + prev[0][k];
                curr[0][k] = max(sell,notsell);
            }
            prev = curr;
        }
        return curr[1][K];
    }
};

File: GFG/DP/38.Buy-ans-Sell-Stocks-IV.java
// Maximum Profit (K transaction)

class Solution {
    int[][][] dp;

    public int solve(int ind, int n, int canBuy, int transactions, int[] prices) {
        if (transactions < 0) return 0;
        if (ind == n) return 0;

        if (dp[ind][transactions][canBuy] != -1) return dp[ind][transactions][canBuy];

        int profit = solve(ind + 1, n, canBuy, transactions, prices);
        if (canBuy == 1) {
            profit = Math.max(profit, -prices[ind] + solve(ind + 1, n, 0, transactions - 1, prices));
        }
        else {
            profit = Math.max(profit, +prices[ind] + solve(ind + 1, n, 1, transactions, prices));
        }

        return dp[ind][transactions][canBuy] = profit;
    }

    public int maxProfit(int k, int[] prices) {
        int n = prices.length;

        dp = new int[n][k + 1][2];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= k; j++) {
                Arrays.fill(dp[i][j], -1);
            }
        }

        return solve(0, n, 1, k, prices);
    }
}
